pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,279 @@
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	"github.com/openshift/origin/pkg/controller/shared"
+)
+
+func init() {
+	admission.RegisterPlugin("RestrictSubjectBindings",
+		func(kclient kclientset.Interface, config io.Reader) (admission.Interface,
+			error) {
+			return NewRestrictUsersAdmission(kclient)
+		})
+}
+
+// restrictUsersAdmission implements admission.Interface and enforces
+// restrictions on adding role bindings in a project to permit only designated
+// collaborators to add role bindings to a project.
+type restrictUsersAdmission struct {
+	*admission.Handler
+
+	oclient             oclient.Interface
+	kclient             kclientset.Interface
+	policyBindingLister oclient.PolicyBindingsListerNamespacer
+	policyBindingSynced func() bool
+}
+
+var _ = oadmission.WantsOpenshiftClient(&restrictUsersAdmission{})
+var _ = oadmission.WantsInformers(&restrictUsersAdmission{})
+var _ = oadmission.Validator(&restrictUsersAdmission{})
+
+const (
+	timeToWaitForCacheSync = 10 * time.Second
+)
+
+// NewRestrictUsersAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project to permit only designated
+// collaborators to add role bindings to a project.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// the project must be annotated to indicate that the subject of the role
+// binding is a collaborator on the project.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
+		return nil
+	}
+
+	// Ignore all operations that correspond to subresource actions.
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	ns := a.GetNamespace()
+
+	// Ignore cluster-level resources.
+	if len(ns) == 0 {
+		return nil
+	}
+
+	binding, ok := a.GetObject().(*authorizationapi.RoleBinding)
+	if !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type: %t", a.GetObject()))
+	}
+
+	if !q.waitForSyncedStores(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(a, errors.New("caches not synchronized"))
+	}
+
+	glog.V(4).Infof("Handling rolebinding %s/%s", binding.Namespace, binding.Name)
+
+	// Only bother checking deltas.
+	newSubjects := []kapi.ObjectReference{}
+	policy := authorizationapi.GetPolicyBindingName(binding.RoleRef.Namespace)
+	policybinding, _ := q.policyBindingLister.PolicyBindings(ns).Get(policy)
```

Comment:
The old object is provided via `a.GetOldObject()`.  Use that instead of a cached. copy.